lib/_http_outgoing.js

```diff
@@ -109,35 +109,29 @@ OutgoingMessage.prototype.setTimeout = function(msecs, callback) {
 // any messages, before ever calling this.  In that case, just skip
 // it, since something else is destroying this connection anyway.
 OutgoingMessage.prototype.destroy = function(error) {
-  if (this.socket)
+  if (this.socket) {
+    // If the socket is corked from a write, uncork it before destroying it.
+    if (this.socket._writableState.corked)
+      this.socket.uncork();
```

Comment:
> > IMHO not transmitting the data when destroy() is hit is the correct behavior of destroy().
> Yes, that's how it behaves right now. .write() can however flush the data synchronously if the socket is ready, so this has undefined behavior:

```js
res.write('this may or may not send');
res.destroy();
```

The behavior of `destroy()` is something we should define and clarify throughout core.
I'm ok with this particular change. Can you add a reference to the failing unit test in the comment? Otherwise it will be extremely hard to make the connection when looking at the code.


> > Beware that, if the socket was corked twice, we need to call enough uncork() for this to take into effect.
> That's not what I'm doing here — we add only a single cork, so we uncork only once. If the user wants to cork the socket, write to it and destroy it, we shouldn't flush the message. That would be backwards-incompatible.

In this PR, every time `write()` is called, `this.connection.cork()` is called. It is entirely possible that `_readableState.corked` is greater that one.

```js
res.write('a')
res.write('b')
res.destroy() // will not uncork
```